### PR TITLE
feat(ci): Add CI pipeline and pre-commit hooks

### LINKED ISSUE #6: feat(ci): Add CI pipeline and linting to CinderPeak.
### Current Situation:
- CinderPeak is a C++ graph library for handling directed, undirected, and mixed graphs, but the repository currently lacks basic documentation and development tooling :contentReference[oaicite:1]{index=1}.
- No README, contribution guide, CI workflows, or test suite are visible—this makes onboarding and collaboration difficult.

### Proposed Enhancements:

1. **Add README.md**
   - Cover project purpose, supported graph types, installation/compilation steps, example usage, and contribution guidelines.

2. **Set Up Continuous Integration (CI):**
   - Use GitHub Actions to compile library across supported platforms (Linux/macOS/Windows).
   - Run static analysis (e.g. clang-tidy) and build verification on each PR.

3. **Introduce Linting & Formatting:**
   - Use tools like clang-format or cpplint to enforce consistent code style.
   - Add configuration files and enable CI checks to ensure formatting compliance.

4. **Add Unit Tests & Testing Framework:**
   - Integrate a testing framework like Google Test or Catch2.
   - Add at least 2–3 sample test cases for basic graph operations (node/edge insertion, traversal).

5. **Developer Documentation:**
   - Provide instructions on building, testing, code style, and contributing.

### Expected Outcomes:
- The project becomes much **more approachable for new contributors**.
- Automated CI ensures the code remains **robust, portable, and maintainable**.
- A clear foundation for future enhancements and modular feature development.

### Suggested Tech Stack:
- CI: GitHub Actions
- Formatting: clang-format or cpplint
- Tests: Google Test or Catch2
- Build: CMake or Makefile (if not already present)

### Difficulty Level:
**Hard** – requires knowledge of C++ build systems, CI/CD, code style tools, and testing frameworks.

### Why This Issue Is Unique:
- No obvious issue currently exists for foundational tooling or documentation in this repo :contentReference[oaicite:2]{index=2}.
- Provides developer ecosystem improvements at a critical stage of project maturity.
- Signals long-term thinking and readiness to maintain project quality alongside functional code.

---

### 🙋 Ready to Contribute  
I’m fully prepared to handle the entire scope:
- Draft and add README,
- Configure CI workflows (lint/build/test),
- Integrate formatting enforcement,
- Set up test framework with sample unit tests,
- Write contribution and developer docs.

If this aligns with your goals, please assign the issue to me—I’d be excited to get started! 🚀

---

I can also prepare a **short version** (concise title + bullet summary) ready for posting, or a comment draft to request assignment—just let me know!

### PR DESCRIPTION
**Pull request by thedgarg31**
✅ Clean CI Pipeline & Pre-commit Hook Integration
This PR introduces a clean, focused implementation of the requested CI and pre-commit infrastructure, as per Anirudh’s instructions.

🔧 What's Included:
Pre-commit Hooks (Linux/macOS & Windows)

scripts/install-hooks.sh and install-hooks.bat

Runs clang-format & clang-tidy on staged C++ files

CI Build Workflows

.github/workflows/build.yml

Separate jobs for Linux, macOS, and Windows (uses Ninja)

Config: BUILD_TESTS=OFF, BUILD_EXAMPLES=OFF

Docs

DEVELOPMENT.md with setup and usage instructions

🚫 What's Not Touched:
No changes to source, test, deprecated files, or CMakeLists.txt

✅ The branch is ready for testing: feat/ci-pipeline-clean
Let's verify the Windows build and proceed to finalize!

### CODE REVIEW

**Comment by SharonIV0x86** [comment]
@thedgarg31 Alright, the workflows seem to work and the formatting also works on linux with commits, but i am not sure about the windows.
You need to rename the ``DEVELOPMENT.md`` to ``development.md`` and move this file to the ``/docs`` folder.

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    - name: Checkout code
+      uses: actions/checkout@v4
+
+    - name: Setup CMake
+      uses: jwlawson/actions-setup-cmake@v1
+      with:
+        cmake-version: '3.20'
+
+    - name: Configure CMake
+      run: cmake -S . -B build -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=OFF -DBUILD_EXAMPLES=OFF
```
Comment (SharonIV0x86) [inline .github/workflows/build.yml]:
```suggestion
      run: cmake -S . -B build -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=ON -DBUILD_EXAMPLES=ON
```

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
```
Comment (SharonIV0x86) [inline .github/workflows/build.yml]:
```suggestion
      run: cmake -S . -B build -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=ON -DBUILD_EXAMPLES=ON
```

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        cmake-version: '3.20'
+
+    - name: Setup Ninja
+      uses: seanmiddleditch/gha-setup-ninja@master
+
+    - name: Setup MSVC
+      uses: ilammy/msvc-dev-cmd@v1
+
+    - name: Configure CMake
+      run: cmake -S . -B build -G Ninja -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=OFF -DBUILD_EXAMPLES=OFF
```
Comment (SharonIV0x86) [inline .github/workflows/build.yml]:
```suggestion
      run: cmake -S . -B build -G Ninja -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTS=ON -DBUILD_EXAMPLES=ON
```

**Review by SharonIV0x86 — CHANGES_REQUESTED** [review]
Need changes.

**Comment by thedgarg31** [comment]
I’ve implemented all the requested changes, including renaming and moving DEVELOPMENT.md to docs/development.md. However, after the updates, the CI workflows for macOS, Linux, and Windows are now failing. Could you please help review the build logs so we can identify and fix the root cause?

**Comment by SharonIV0x86** [comment]
@thedgarg31 All the workflows seem to work as expected. Check twice.

**Comment by thedgarg31** [comment]
"Thanks for confirming the workflows are working fine. I’ve double-checked and all checks are passing now. Ready for merge from my side ✅"

**Review by SharonIV0x86 — APPROVED** [review]
Approved.
Will be merged in sometime. Till then if you want you can work on some new issues.

**Comment by thedgarg31** [comment]
Sure.Thanku

**Comment by SharonIV0x86** [comment]
@thedgarg31 
To assign your points for GSSOC, I’ll need your full name (or the exact name you registered with for GSSOC). Please ensure the information is accurate, as incorrect details may result in your points not being credited.

**Comment by thedgarg31** [comment]
Ya Sure .
I registered with my full name - DAKSH GARG 
Thanks

**Comment by SharonIV0x86** [comment]
@thedgarg31 Thank you for contributing, your PR has been merged.
Kindly ⭐ the repository as it shows appreciation to the maintainers.

**Comment by thedgarg31** [comment]
Done

**Comment by thedgarg31** [comment]
Thanks for merging my PR and assigning the GSSoC points. Could you please let me know when they will be reflected on the leaderboard and where I can check them?

**Comment by SharonIV0x86** [comment]
> Thanks for merging my PR and assigning the GSSoC points. Could you please let me know when they will be reflected on the leaderboard and where I can check them?

There is a lot of confusion with the official leaderboard and the whole points assigning process is manual, like PAs have to manually keep track of the points in a google sheet and submit that sheet to the managers whenever requested, so its nothing in our hands it can take a week or so to get your points reflected on the leaderboard once it is live but i am not sure.

**Comment by SharonIV0x86** [comment]
@thedgarg31 You can check your points here
https://gssoc25-leaderboard.vercel.app/download
search by the project.

**Comment by thedgarg31** [comment]
Thanku so much
